README.md
# Toolbar Extras for Oxygen Builder - Power Up Your Admin Bar 

[![Plugin Version](https://img.shields.io/wordpress/plugin/v/toolbar-extras-oxygen.svg)](https://wordpress.org/plugins/toolbar-extras-oxygen/)
[![WordPress Version Compatibility](https://img.shields.io/wordpress/v/toolbar-extras-oxygen.svg)](https://wordpress.org/plugins/toolbar-extras-oxygen/)
[![Downloads](https://img.shields.io/wordpress/plugin/dt/toolbar-extras-oxygen.svg)](https://wordpress.org/plugins/toolbar-extras-oxygen/)
[![Rating](https://img.shields.io/wordpress/plugin/r/toolbar-extras-oxygen.svg)](https://wordpress.org/plugins/toolbar-extras-oxygen/)

**Contributors:** daveshine, deckerweb, wpautobahn, toolbarextras  
**Donate link:** https://www.paypal.me/deckerweb  
**Tags:** toolbar, adminbar, admin bar, oxygen, page builder, website, site builder, deckerweb, ddwtoolbar  
**Requires at least:** 4.7  
**Tested up to:** 5.2  
**Requires PHP:** 5.6  
**Stable tag:** 1.2.0  
**License:** GPL-2.0-or-later  
**License URI:** https://opensource.org/licenses/GPL-2.0  

Building websites with Oxygen? Your work will get easier & faster with this Toolbar Extras Add-On. Extended plugin support baked right in.

[<img src="https://raw.githubusercontent.com/deckerweb/toolbar-extras-oxygen/master/assets-repos/github-com/tbexob-banner.png" data-canonical-src="https://raw.githubusercontent.com/deckerweb/toolbar-extras-oxygen/master/assets-repos/github-com/tbexob-banner.png" width="772" height="250" />](https://toolbarextras.com/)


## Description 

🚀 Quick Access to Your Oxygen Site Building Resources - Big Time Saver. Perfect for Non-Techies and Non-Coder Designers.  
This smart WordPress plugin just adds some missed and **very useful Oxygen-specific admin links** to your Toolbar / Admin Bar. Save scrolling travel and time. Access settings and content pages faster from one central area in the Admin Dashboard AND when on the frontend of your site. Plus few optimizations for Oxygen's own Toolbar within the Builder :-).

[![Video of Plugin's Live Demo and Walkthrough](https://img.youtube.com/vi/jmvFqH9pqrA/0.jpg)](https://www.youtube.com/watch?v=jmvFqH9pqrA)  
[**original video link**](https://www.youtube.com/watch?v=jmvFqH9pqrA) *by plugin developer David Decker*

ℹ️ **Essential:** Toolbar Extras for Oxygen Builder is an **Add-On** plugin, it **requires** the base plugin "Toolbar Extras".  
ℹ️ **Important:** "Oxygen Builder" is a **premium** plugin for WordPress by Soflyy.  
ℹ️ **Disclaimer:** Toolbar Extras for Oxygen Builder is an **independent** free plugin by David Decker from DECKERWEB. It does **not** represent Soflyy/ Oxygen.


### ♥️ Features - What the Plugin Does 
* **More links for the "Back to WP" section** in Oxygen Builder interface itself -- Oxygen Templates, WordPress Pages, WooCommerce Products, Oxygen User Elements Library
* Optionally open those **"Back to WP" links in a new tab/ window** -- perfect for **previewing** your templates/ pages ...
* Useful **Row action "Edit with Oxygen"** for all Oxygen Template post types, plus any item of an Oxygen-enabled post type that contains an Oxygen Element already
* Useful **Post State "Oxygen"** appended to all post type items edited with Oxygen - for example for Pages
* Registers Oxygen Builder - **set as default Builder** for the "Build" Group
* Adds Oxygen Templates **links to the "Build" Group** in the Toolbar, including Reusable Parts and User Elements Library (2.3+)
* Optionally adds list of Oxygen Templates to the Toolbar as extra group
* Optionally adds list of Oxygen-edited Pages to the Toolbar as extra group
* Label tweaks and additions for Oxygen Templates in the "New Content" Group
* Adds Website install wizard/ Importer to the Toolbar "Build" Group as well - for faster access
* Adds all Oxygen settings and settings tabs links to the Toolbar
* Adds links to settings of Oxygen Add-On Plugins (only if these are active)
* Adds lots of official and Community resources to the Toolbar (optional, can be disabled via settings) - documentation, forum, Facebook Group etc.
* Frontend only: The original Oxygen Toolbar item can be customized - icon, location, label
* Add additional view to filter Oxygen Reusable Parts only - in Templates post type list table (overview table)
* Proper highlighting of Templates submenu in Oxygen's admin menu - so you always know where you are
* Adds new tab on plugin's settings page (under "Settings > Toolbar Extras > Tab: Oxygen") so you have control over what is displayed and what's not!
* Corrected post type labels for Oxygen Templates ('ct_template') and User Elements Library ('oxy_user_library') where needed
* Use any other "global" Toolbar Extras feature like automatic detection of Local Development Enviroment, Dev Mode, and lots of customization options


### 👍 More Benefits 
* Any items are only loaded if needed
* Perfect integration into Toolbar Extras' items and settings (of course)
* Non-Techie and Non-Coder Designer friendly :)
* At the same time very developer friendly as well - offering quite a few action hooks and filters for customization if needed ;-)
* Clean and efficient code
* Fully internationalized and translateable - also tested/optimized for RTL languages


### 🔌 Oxygen Add-On Support/ Integration 
* *Official:* **Oxygen Gutenberg Integration** (Premium, by Soflyy) -- For the Block Editor
* *Official:* **Oxygen Elements for WooCommerce** (Premium, by Soflyy)
* [**Oxygens Swiss Knife** (free, by Marko Krstic)](https://github.com/krstivoja/Oxygens-Swiss-Knife)
* [**Oxygen Theme Enabler** (free, by Sridhar Katakam)](https://github.com/srikat/oxygenthemeenabler) -- **Note:** The settings of this Add-On will be considered for Toolbar items! ;-)
* [**Builder Template Categories** (free, by David Decker - DECKERWEB)](https://wordpress.org/plugins/builder-template-categories/)
* [**My Custom Functionality** (free, by Sridhar Katakam)](https://github.com/srikat/my-custom-functionality)
* [**Oxygen Block Lab Support** (free, by David Browne)](https://github.com/wplit/oxygen-block-lab-support)
* [**Oxygen EDD Support** (free, by David Browne)](https://github.com/wplit/Oxygen-EDD-Support)
* [**Oxygen RCP Support** (free, by David Browne)](https://github.com/wplit/Oxygen-RCP-Support)


### 👉 A Typical Workflow / Usage 
* Install Oxygen Builder, Toolbar Extras, plus the Toolbar Extras Oxygen Add-On - in that order
* In Toolbar Extras settings set Oxygen as your default "Page Builder", set optional background color of Toolbar etc.
* In the Oxygen-specific settings enable or disable various tweaks regarding Oxygen in WP-Admin, the Toolbar, and the Builder itself
* When working on your site(s) focus on the Toolbar in the admin and on the frontend to get a habit for this "central hub"


### ℹ️ Important: Required/ Recommended for plugin usage 
* Required: User has role `Administrator` (needed capabilities `manage_options` and `edit_theme_options`)
* Required: Toolbar/ Admin Bar is enabled (user profile settings in WordPress)
* Required: User is logged in (of course)
* Required: The premium plugin "Oxygen Builder" (version 2.0 or higher) is installed and activated (otherwise this Add-On will do nothing!)
* Recommended: Plugin's settings are enabled to show the wanted links (or any filter/ constant in code)


### 🌎 Translations 
* 🇺🇸 English (United States) - `en_US` = default, always included
* 🇩🇪 [German (informal, default)](https://translate.wordpress.org/locale/de/default/wp-plugins/toolbar-extras-oxygen) - `de_DE`
* 🇩🇪 [German (formal)](https://translate.wordpress.org/locale/de/formal/wp-plugins/toolbar-extras-oxygen) - `de_DE_formal`
* `.pot` file (`toolbar-extras-oxygen.pot`) for translators is always included in the plugin's 'languages' folder :)


### 😎 Be a Contributor 
If you want to translate, [go to the Translation Portal at translate.wordpress.org](https://translate.wordpress.org/projects/wp-plugins/toolbar-extras-oxygen).

You can also contribute code-wise via our [Toolbar Extras for Oxygen Builder GitHub Repository](https://github.com/deckerweb/toolbar-extras-oxygen) - and see where you can help.


### 📝 Documentation and Support 
* For more information about features, FAQs and documentation, check out our website at [Toolbar Extras for Oxygen Builder Plugin](https://toolbarextras.com/).
* If you have any more questions, visit our support on the [Plugin's Forum](https://wordpress.org/support/plugin/toolbar-extras-oxygen).
* An overview about the ongoing development you'll find on our [Trello Board for the Plugin's Roadmap](https://trello.com/b/JrpjwlX4/toolbar-extras-public-roadmap)


### ⚡ Liked Toolbar Extras for Oxygen Builder Plugin? 
* Join our [Facebook Community Group 💬](https://www.facebook.com/groups/ToolbarExtras/)
* **Rate us 5 ⭐ stars** on [WordPress.org](https://wordpress.org/support/plugin/toolbar-extras-oxygen/reviews/?filter=5/#new-post) :)
* [**Subscribe to my Newsletter for insider info on this plugin** 💯](https://eepurl.com/gbAUUn), plus tutorials and more stuff on deckerweb WordPress plugins - join a thriving community of site builders!
* [**Become a Patron** 💜](https://www.patreon.com/deckerweb) and support ongoing development, maintenance and support of this plugin
* [**Follow us on Instagram** 📷](https://www.instagram.com/toolbarextras) and like our content
* Follow us on Twitter 🐦: [@ToolbarExtras](https://twitter.com/toolbarextras) and [@deckerweb](https://twitter.com/deckerweb)


### ☕ This Plugin ... 
* ... scratches my own itch!
* ... is *Quality Made in Germany*
* ... was created with love (plus some coffee) :-) - [if you like it consider donating](https://www.paypal.me/deckerweb)



## Installation 


### Minimum Requirements 

* **Base plugin:** [Toolbar Extras](https://wordpress.org/plugins/toolbar-extras/), version 1.4.6 or higher
* WordPress version 4.7 or higher
* PHP version 5.6.20 or higher
* MySQL version 5.0 or higher
* Administrator user with capabilities `manage_options` and `edit_theme_options` who is logged-in and has the Toolbar (Admin Bar) enabled (user profile)


### We Recommend Your Host Supports at least: 

* PHP version 7.2 or higher
* MySQL version 5.6 or higher / OR MariaDB 10.1 or higher
* HTTPS support


### Installation 

[![Video of how to Install & Activate this Add-On and set the first important setting - Live Walkthrough](https://img.youtube.com/vi/NrYj72Lpdww/0.jpg)](https://www.youtube.com/watch?v=NrYj72Lpdww)  
[**original video link**](https://www.youtube.com/watch?v=NrYj72Lpdww) *by plugin developer David Decker*

1. Install using the WordPress built-in Plugin installer (via **Plugins > Add New** - search for `toolbar extras oxygen`), or extract the ZIP file and drop the contents in the `wp-content/plugins/` directory of your WordPress installation.
2. Activate the plugin through the 'Plugins' menu in WordPress.
3. Optional: if Toolbar Extras base plugin isn't installed/activated yet, just follow the info and install/activate button in the admin notice that will appear
4. Go to **Settings > Toolbar Extras > General** to set "Oxygen Builder" as your **default** Page Builder
5. Now enjoy building your site with Oxygen Builder - and let the Toolbar work for you ;-)

For documentation and tutorials visit our [Knowledge Base](https://toolbarextras.com/docs/).



## Frequently Asked Questions 


### Is Oxygen Builder required for this plugin? 
Technically Oxygen Builder is not required. However, without Oxygen Builder being activated this Add-On plugin (Toolbar Extras for Oxygen Builder) doesn't make any sense at all.

I totally [recommend Oxygen Builder](https://toolbarextras.com/go/oxygen/) - so, with it active, this plugin here will make the perfect sense, helping you save time, scrolling travel and getting you more productive.



### Will other third-party Oxygen Plugins be supported? 
Yes, absolutely. - Every Oxygen-specific plugin or Add-On will be supported as long as it has a settings page or comes with a post type screen or similar.



### Why is the Administrator Role required? 
This Add-On plugin only works and makes sense for Administrator users (and Super Admins in Multisite) just because the links it adds to the Toolbar normally do require those permissions for the original supported plugins.

Example: The Oxygen Builder settings page requires the capability `manage_options` - by default, only Administrators have this capability. So Toolbar Extras for Oxygen Builder "inherits" those capabilities (permissions) as well.

So, to see anything added to the Toolbar, it is required the user must have the capabilities of `manage_options` and `edit_theme_options`. As long as these are in place, Toolbar Extras is the perfect companion! :-)

(Note: Implementing other roles/ capabilities is not planned, and it makes no sense, really. It would require the codebase to be bloated with all those additional permission conditions and would give only very very few (if any at all!) additional links for roles like "Editor" or lower. Again, offering all items only for Admins follows the supported integrations, prevents bloat and big footprint while also avoiding very high maintenance costs.)



### Will this Plugin slow down my site? 
**Absolutely not.** The plugin will only do its stuff for logged-in Administrator users. It does nothing for visitors of your site. And, the plugin loads its stuff only when needed and if supported third-party plugins are active. Plugin was built to be as lightweight as possible.



### How can I disable the link attributes (tooltips)? 
This is easily possible within the "General" settings of Toolbar Extras. Just go to the "General" tab and set options for the link behavior. -- Note: This setting also applies to the "Back to WP" section in Oxygen Builder interface itself.

Settings > Toolbar Extras > Tab: "General" > Section: "Links Behavior"



### Where are more settings? 
A lot more settings are available via the base plugin, Toolbar Extras. This opens up a lot of customization options and tweaks for you.



### Does the Plugin work with the new Block Editor (Gutenberg)? 

CHANGES.md
# Toolbar Extras for Oxygen Builder (Add-On)

**General Info**

* [Plugin's own website: toolbarextras.com](https://toolbarextras.com/)
* [Plugin page on WordPress.org: wordpress.org/plugins/toolbar-extras/](https://wordpress.org/plugins/toolbar-extras-oxygen/)
* [Translate the plugin](https://translate.wordpress.org/projects/wp-plugins/toolbar-extras-oxygen)
* [**Join plugin's newsletter for insider info, tutorials and more**](https://eepurl.com/gbAUUn)
* [**Become a Patron**](https://www.patreon.com/deckerweb) and support ongoing development, maintenance and support of this plugin
* [**Donate** for the further development & support of the plugin](https://www.paypal.me/deckerweb)
* [Plugin's documentation & knowledge base](https://toolbarextras.com/docs/)
* [Facebook Community User Group](https://www.facebook.com/groups/ToolbarExtras/)

## Changelog of the Plugin

### 🎉 1.2.0 - 2019-09-04

* *New: Successfully tested with Oxygen Builder 3.0, including WooCommerce & Gutenberg Integrations*
* *New: Successfully tested with Oxygen Builder 2.4.1*
* New: For Oxygen Template post type, added new optional **Template Group** as a top-level item, or as sub item in the *Build Group* - has lots of options for customizing
* New: For Oxygen-edited Pages, added new optional **Pages Group** as a top-level item, or as sub item in the *Build Group* - has lots of options for customizing
* New: Added plugin support for [Oxygen Block Lab Support (free, by David Browne)](https://github.com/wplit/oxygen-block-lab-support) - for *Block Lab* block templates
* New: Added plugin support for [Oxygen EDD Support (free, by David Browne)](https://github.com/wplit/Oxygen-EDD-Support) - for *Easy Digital Downloads (EDD)* product & archive templates
* New: Added plugin support for [Oxygen RCP Support (free, by David Browne)](https://github.com/wplit/Oxygen-RCP-Support) - for *Restrict Content Pro* content templates
* New: Added internal logic for settings updater between major versions of this Add-On
* Tweak: Added sub item for optional "Gutenberg" settings (Oxygen Gutenberg Integration)
* Tweak: For "My Custom Functionality" plugin support, added dynamic folder path detection to better handle GitHub specific scenarios
* Tweak: For some existing tweaks made sure they are not executed in the Oxygen Builder context to avoid any errors, warnings and notices
* Tweak: Added [new screenshots (for the above new features)](https://toolbarextras.com/addons/oxygen-builder/#tab-screenshots)
* Tweak: Some under the hood improvements on code base
* Tweak: Updated `.pot` file plus all German translations (formal, informal) and language packs
* Fix: Fixed PHP notice if Oxygen Builder is still set as default page builder in settings but the plugin was deactivated meanwhile
* **Note:** The Add-On version 1.2.0 requires [base plugin Toolbar Extras v1.4.6+](https://wordpress.org/plugins/toolbar-extras/)


### 🎉 1.1.0 - 2019-05-16

* *New: Successfully tested with Oxygen Builder 2.3*
* *New: Successfully tested with WordPress 5.2*
* New: Added plugin support for ["My Custom Functionality" (free, by Sridhar Katakam)](https://github.com/srikat/my-custom-functionality)
* New: Integrated with WordPress 5.2+ new Site Health feature: Toolbar Extras for Oxygen Builder now has an extra section on the Debug Info tab - this is especially helpful for support requests
* Tweak: Greatly improved the smart tweak for proper highlighting of parent file/ submenu for Oxygen Templates and User Elements Library (Block Library) as well
* Tweak: Improved all tweaks regarding Oxygen Templates post type labels
* Tweak: Plugin Manager admin notices are now properly dismissed - switched to notice library of Toolbar Extras base plugin :) 
* Tweak: Minor compatibility tweaks for Oxygen 2.3 beta/RC/final releases
* Tweak: Updated `.pot` file plus all German translations (formal, informal) and language packs
* **Note:** The Add-On version 1.1.0 requires [base plugin Toolbar Extras v1.4.3+](https://wordpress.org/plugins/toolbar-extras/)


### ⚡ 1.0.2 - 2019-04-18

* Tweak: Setup Plugin Manager to load earlier to actually appear again
* Tweak: Small improvement to User Group link title attribute
* Tweak: Internal improvements and optimizations
* Tweak: Updated `.pot` file plus all German translations (formal, informal) and language packs
* Tweak: Enhanced and improved Readme.txt file here - also added new FAQ entry
* Fix: Proper Editor URL query params for templates to make all editable again - thanks to user feedback in [Oxygen FB Group](https://www.facebook.com/groups/1626639680763454/)
* Fix: Correct admin URL on Plugins page listing


### ⚡ 1.0.1 - 2019-04-02

* New: [Video of plugin walkthrough and live demo](https://www.youtube.com/watch?v=jmvFqH9pqrA)
* New: [Video of Add-On installation & activation, plus setting the first important setting](https://www.youtube.com/watch?v=NrYj72Lpdww)
* Tweak: Further performance optimization to load as minimal stuff as possible
* Tweak: Updated `.pot` file plus all German translations (formal, informal) and language packs
* Fix: Fixed fatal error on activation (sorry!)


### 🎉 1.0.0 - 2019-04-02

* *Plugin launch. Everything's new!*
* New: Including support for 2 Oxygen-specific Add-On plugins
* New: New "Oxygen" tab on Toolbar Extras plugin settings page, with up to 16 different options
* New: Few tweaks for "Back to WP" links within Oxygen's own Toolbar in the Builder
* New: Post type item *Row Action* for "Edit with Oxygen"
* New: Post type item *Post State* for "Oxygen"
* New: 12 great Oxygen resources links integrated into the Toolbar
* *Note: the Add-On integrates with current Oxygen Builder v2.2.1 as well as the upcoming v2.3.x release (tested with 2.3 alpha 1)*


### ⚡ 0.9.0 - 2019-03-27

* New: Private beta release on GitHub


includes/admin/admin-extras.php
<?php

// includes/admin/admin-extras

/**
 * Prevent direct access to this file.
 *
 * @since 1.0.0
 */
if ( ! defined( 'ABSPATH' ) ) {
	exit( 'Sorry, you are not allowed to access this file directly.' );
}


//


//


/**
 * Add "Settings" and Custom Menu" links to Plugins page.
 *
 * @since 1.0.0
 * @since 1.2.0 Overall code improvements.
 *
 * @uses ddw_tbexob_is_toolbar_extras_active()
 * @uses ddw_tbexob_is_oxygen_active()
 *
 * @param array $action_links (Default) Array of plugin action links.
 * @return array Modified array of plugin action links.
 */
function ddw_tbexob_custom_settings_links( $action_links = [] ) {

	$tbexob_links = array();

	/** Add settings link only if user can 'manage_options' */
	if ( current_user_can( 'manage_options' ) ) {

		/** If environment is not ready point to plugin manager */
		if ( ( ddw_tbexob_is_toolbar_extras_active() && version_compare( TBEX_PLUGIN_VERSION, TBEXOB_REQUIRED_BASE_PLUGIN_VERSION, '>=' ) )
			&& ! ddw_tbexob_is_oxygen_active()
		) {

			$tbexob_links[ 'settings' ] = sprintf(
				'<a href="%s" title="%s"><span class="dashicons-before dashicons-admin-plugins"></span> %s</a>',
				esc_url( admin_url( 'plugins.php?page=toolbar-extras-suggested-plugins' ) ),
				esc_html__( 'First Step: Setup Environment to use the plugin', 'toolbar-extras-oxygen' ),
				esc_attr__( 'First Step: Setup Environment', 'toolbar-extras-oxygen' )
			);

		}  // end if

		/** Oxygen & Settings Page links */
		if ( ddw_tbexob_is_toolbar_extras_active() && ddw_tbexob_is_oxygen_active() ) {

			$tbexob_links[ 'oxygen' ] = sprintf(
				'<a href="%s" title="%s"><span class="dashicons-before dashicons-welcome-widgets-menus"></span> %s</a>',
				esc_url( admin_url( 'admin.php?page=ct_dashboard_page' ) ),
				esc_html__( 'Oxygen Builder', 'toolbar-extras-oxygen' ),
				esc_attr__( 'Oxygen', 'toolbar-extras-oxygen' )
			);

			$tbexob_links[ 'settings' ] = sprintf(
				'<a href="%s" title="%s"><span class="dashicons-before dashicons-admin-generic"></span> %s</a>',
				esc_url( admin_url( 'options-general.php?page=toolbar-extras&tab=oxygen' ) ),
				esc_html__( 'Toolbar Settings for Oxygen Builder', 'toolbar-extras-oxygen' ),
				esc_attr__( 'Toolbar Settings', 'toolbar-extras-oxygen' )
			);

		}  // end if

	}  // end if

	/** Display plugin settings links */
	return apply_filters(
		'tbexob/filter/plugins_page/settings_link',
		array_merge( $tbexob_links, $action_links ),
		$tbexob_links		// additional param
	);

}  // end function


add_filter( 'plugin_row_meta', 'ddw_tbexob_plugin_links', 10, 2 );
/**
 * Add various support links to Plugins page.
 *
 * @since 1.0.0
 *
 * @uses ddw_tbex_get_info_link()
 *
 * @param array  $tbexob_links (Default) Array of plugin meta links
 * @param string $tbexob_file  Path of base plugin file
 * @return array $tbexob_links Array of plugin link strings to build HTML markup.
 */
function ddw_tbexob_plugin_links( $tbexob_links, $tbexob_file ) {

	/** Capability check */
	if ( ! current_user_can( 'install_plugins' ) ) {
		return $tbexob_links;
	}

	/** List additional links only for this plugin */
	if ( $tbexob_file === TBEXOB_PLUGIN_BASEDIR . 'toolbar-extras-oxygen.php' ) {

		?>
			<style type="text/css">
				tr[data-plugin="<?php echo $tbexob_file; ?>"] .plugin-version-author-uri a.dashicons-before:before {
					font-size: 17px;
					margin-right: 2px;
					opacity: .85;
					vertical-align: sub;
				}
			</style>
		<?php

		/* translators: Plugins page listing */
		$tbexob_links[] = ddw_tbex_get_info_link(
			'url_wporg_forum',
			esc_html_x( 'Support', 'Plugins page listing', 'toolbar-extras-oxygen' ),
			'dashicons-before dashicons-sos',
			'oxygen'
		);

		/* translators: Plugins page listing */
		$tbexob_links[] = ddw_tbex_get_info_link(
			'url_fb_group',
			esc_html_x( 'Facebook Group', 'Plugins page listing', 'toolbar-extras-oxygen' ),
			'dashicons-before dashicons-facebook'
		);

		/* translators: Plugins page listing */
		$tbexob_links[] = ddw_tbex_get_info_link(
			'url_translate',
			esc_html_x( 'Translations', 'Plugins page listing', 'toolbar-extras-oxygen' ),
			'dashicons-before dashicons-translation',
			'oxygen'
		);

		/* translators: Plugins page listing */
		$tbexob_links[] = ddw_tbex_get_info_link(
			'url_donate',
			esc_html_x( 'Donate', 'Plugins page listing', 'toolbar-extras-oxygen' ),
			'button dashicons-before dashicons-thumbs-up'
		);

		/* translators: Plugins page listing */
		$tbexob_links[] = ddw_tbex_get_info_link(
			'url_newsletter',
			esc_html_x( 'Join our Newsletter', 'Plugins page listing', 'toolbar-extras-oxygen' ),
			'button-primary dashicons-before dashicons-awards'
		);

	}  // end if plugin links

	/** Output the links */
	return apply_filters(
		'tbexob/filter/plugins_page/more_links',
		$tbexob_links
	);

}  // end function


add_filter( 'admin_footer_text', 'ddw_tbexob_admin_footer_tweak' );
/**
 * On the "Oxygen" settings tab add footer text to invite for plugin review.
 *
 * @since 1.0.0
 *
 * @uses get_current_screen()
 *
 * @param string $footer_text The content that will be printed.
 * @return string The content that will be printed.
 */
function ddw_tbexob_admin_footer_tweak( $footer_text ) {

	/** Current screen logic */
	$current_screen = get_current_screen();
	$is_tbex_screen = array(
		'plugins_page_toolbar-extras-oxygen-suggested-plugins',
	);

	/** Active settings tab logic */
	$active_tab = isset( $_GET[ 'tab' ] ) ? sanitize_key( wp_unslash( $_GET[ 'tab' ] ) ) : '';

	/** Conditionally set footer text */
	if ( in_array( $current_screen->id, $is_tbex_screen )
		|| ( 'settings_page_toolbar-extras' === $current_screen->id && 'oxygen' === $active_tab )
	) {

		$rating = sprintf(
			/* translators: %s - 5 stars icons */
			'<a href="https://wordpress.org/support/plugin/toolbar-extras-oxygen/reviews/?filter=5#new-post" target="_blank" rel="noopener noreferrer">' . __( '%s rating', 'toolbar-extras-oxygen' ) . '</a>',
			'&#9733;&#9733;&#9733;&#9733;&#9733;'
		);

		$footer_text = sprintf(
			/* translators: 1 - Toolbar Extras for Oxygen Builder / 2 - label "5 star rating" */
			__( 'Enjoyed %1$s? Please leave us a %2$s. We really appreciate your support!', 'toolbar-extras-oxygen' ),
			'<strong>' . __( 'Toolbar Extras for Oxygen Builder', 'toolbar-extras-oxygen' ) . '</strong>',
			$rating
		);

	}  // end if

	/** Render footer text */
	return $footer_text;

}  // end function


add_filter( 'debug_information', 'ddw_tbexob_site_health_add_debug_info', 6 );
/**
 * Add additional plugin related info to the Site Health Debug Info section.
 *   (Only relevant for WordPress 5.2 or higher)
 *
 * @link https://make.wordpress.org/core/2019/04/25/site-health-check-in-5-2/
 *
 * @since 1.1.0
 * @since 1.2.0 Added new options; tweaks & improvements.
 *
 * @uses ddw_tbex_string_debug_diagnostic()
 * @uses ddw_tbex_string_yes()
 * @uses ddw_tbex_string_no()
 * @uses ddw_tbex_string_uninstalled()
 * @uses ddw_tbexob_is_oxygen_user_library_active()
 *
 * @param array $debug_info Array holding all Debug Info items.
 * @return array Modified array of Debug Info.
 */
function ddw_tbexob_site_health_add_debug_info( $debug_info ) {

	/** Add our Debug info */
	$debug_info[ 'toolbar-extras-oxygen' ] = array(
		'label'       => esc_html__( 'Toolbar Extras for Oxygen Builder', 'toolbar-extras-oxygen' ) . ' (' . esc_html__( 'Add-On Plugin', 'toolbar-extras-oxygen' ) . ')',
		'description' => ddw_tbex_string_debug_diagnostic( 'oxygen' ),
		'fields'      => array(

			/** Add-On values etc. */
			'tbexob_plugin_version' => array(
				'label' => __( 'Add-On Plugin version', 'toolbar-extras-oxygen' ),
				'value' => TBEXOB_PLUGIN_VERSION,
			),
			'tbexob_required_base_plugin_version' => array(
				'label' => __( 'Required Base Plugin version', 'toolbar-extras-oxygen' ),
				'value' => TBEXOB_REQUIRED_BASE_PLUGIN_VERSION,
			),
			'tbexob_oxygen_display_customizer' => array(
				'label' => __( 'Display Settings Customizer', 'toolbar-extras-oxygen' ),
				'value' => get_option( 'tbex-options-oxygen', ddw_tbex_string_yes( 'return' ) )[ 'oxygen_display_customizer' ],
			),
			'tbexob_oxygen_tl_priority' => array(
				'label' => __( 'Oxygen item priority', 'toolbar-extras-oxygen' ),
				'value' => get_option( 'tbex-options-oxygen', '1000' )[ 'oxygen_tl_priority' ],
			),
			'tbexob_oxygen_row_actions' => array(
				'label' => __( 'Use Row Action', 'toolbar-extras-oxygen' ),
				'value' => get_option( 'tbex-options-oxygen', ddw_tbex_string_yes( 'return' ) )[ 'oxygen_row_actions' ],
			),
			'tbexob_oxygen_post_state' => array(
				'label' => __( 'Use Post State', 'toolbar-extras-oxygen' ),
				'value' => get_option( 'tbex-options-oxygen', ddw_tbex_string_yes( 'return' ) )[ 'oxygen_post_state' ],
			),
			'tbexob_oxygen_btwp_links' => array(
				'label' => __( 'Additional links Back to WP section', 'toolbar-extras-oxygen' ),
				'value' => get_option( 'tbex-options-oxygen', ddw_tbex_string_yes( 'return' ) )[ 'oxygen_btwp_links' ],
			),
			'tbexob_oxygen_btwp_links_blank' => array(
				'label' => __( 'Open additional links in new tab', 'toolbar-extras-oxygen' ),
				'value' => get_option( 'tbex-options-oxygen', ddw_tbex_string_yes( 'return' ) )[ 'oxygen_btwp_links_blank' ],
			),

			/** Oxygen specific */
			'CT_VERSION' => array(
				'label' => 'Oxygen Builder: CT_VERSION',
				'value' => ( ! defined( 'CT_VERSION' ) ? ddw_tbex_string_uninstalled() : CT_VERSION ),
			),
			'tbexob_user_library_active' => array(
				'label' => __( 'User Library active', 'toolbar-extras-oxygen' ),
				'value' => ( ddw_tbexob_is_oxygen_user_library_active() ? ddw_tbex_string_yes( 'return' ) : ddw_tbex_string_no( 'return' ) ),
			),

		),  // end array
	);

	/** Return modified Debug Info array */
	return $debug_info;

}  // end function


/**
 * Inline CSS fix for Plugins page update messages.
 *
 * @since 1.0.0
 *
 * @see ddw_tbex_plugin_update_message()
 * @see ddw_tbex_multisite_subsite_plugin_update_message()
 */
function ddw_tbexob_plugin_update_message_style_tweak() {

	?>
		<style type="text/css">
			.tbexob-update-message p:before,
			.update-message.notice p:empty {
				display: none !important;
			}
		</style>
	<?php

}  // end function


add_action( 'in_plugin_update_message-' . TBEXOB_PLUGIN_BASEDIR . 'toolbar-extras-oxygen.php', 'ddw_tbexob_plugin_update_message', 10, 2 );
/**
 * On Plugins page add visible upgrade/update notice in the overview table.
 *   Note: This action fires for regular single site installs, and for Multisite
 *         installs where the plugin is activated Network-wide.
 *
 * @since 1.0.0
 *
 * @param object $data
 * @param object $response
 * @return string Echoed string and markup for the plugin's upgrade/update
 *                notice.
 */
function ddw_tbexob_plugin_update_message( $data, $response ) {

	if ( isset( $data[ 'upgrade_notice' ] ) ) {

		ddw_tbexob_plugin_update_message_style_tweak();

		printf(
			'<div class="update-message tbexob-update-message">%s</div>',
			wpautop( $data[ 'upgrade_notice' ] )
		);

	}  // end if

}  // end function


add_action( 'after_plugin_row_wp-' . TBEXOB_PLUGIN_BASEDIR . 'toolbar-extras-oxygen.php', 'ddw_tbexob_multisite_subsite_plugin_update_message', 10, 2 );
/**
 * On Plugins page add visible upgrade/update notice in the overview table.
 *   Note: This action fires for Multisite installs where the plugin is
 *         activated on a per site basis.
 *
 * @since 1.0.0
 *
 * @param string $file
 * @param object $plugin
 * @return string Echoed string and markup for the plugin's upgrade/update
 *                notice.
 */
function ddw_tbexob_multisite_subsite_plugin_update_message( $file, $plugin ) {

	if ( is_multisite() && version_compare( $plugin[ 'Version' ], $plugin[ 'new_version' ], '<' ) ) {

		$wp_list_table = _get_list_table( 'WP_Plugins_List_Table' );

		ddw_tbexob_plugin_update_message_style_tweak();

		printf(
			'<tr class="plugin-update-tr"><td colspan="%s" class="plugin-update update-message notice inline notice-warning notice-alt"><div class="update-message tbexob-update-message"><h4 style="margin: 0; font-size: 14px;">%s</h4>%s</div></td></tr>',
			$wp_list_table->get_column_count(),
			$plugin[ 'Name' ],
			wpautop( $plugin[ 'upgrade_notice' ] )
		);

	}  // end if

}  // end function


/**
 * Optionally tweaking Plugin API results to make more useful recommendations to
 *   the user.
 *
 * @since 1.0.0
 */

add_filter( 'ddwlib_plir/filter/plugins', 'ddw_tbexob_register_extra_plugin_recommendations_oxygen' );
/**
 * Register specific plugins for the class "DDWlib Plugin Installer
 *   Recommendations".
 *   Note: The top-level array keys are plugin slugs from the WordPress.org
 *         Plugin Directory.
 *
 * @since 1.0.0
 *
 * @param array $plugins Array holding all plugin recommendations, coming from
 *                       the class and the filter.
 * @return array Filtered and merged array of all plugin recommendations.
 */
function ddw_tbexob_register_extra_plugin_recommendations_oxygen( array $plugins ) {

	/** Remove our own slug when we are already active :) */
	if ( isset( $plugins[ 'toolbar-extras-oxygen' ] ) ) {
		$plugins[ 'toolbar-extras-oxygen' ] = null;
	}

	/** Add new keys to recommendations */
	$plugins[ 'wp-asset-clean-up' ] = array(
		'featured'    => 'yes',
		'recommended' => 'yes',
		'popular'     => 'no',
	);

	/** Return tweaked array of plugins */
	return $plugins;

}  // end function
